### PR TITLE
[Feature] Added support for linting style tags within html files

### LINKED ISSUE #8: <style> tag
Please add support for `<style>` tag in html files. We need for Polymer Theme.

https://github.com/StartPolymer/polymer-starter-kit-plus

### PR DESCRIPTION
**Pull request by blake-newman**
- Added grammar styles for embedded css
- Added option to enable html linting

Fixes #8.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -35,7 +35,7 @@
     "atom-package-deps": "^4.0.1",
     "cosmiconfig": "^1.1.0",
     "deep-assign": "^2.0.0",
-    "stylelint": "5.2.1",
+    "stylelint": "NEEDS UPDATING WHEN AVAILABLE",
```
Comment (Arcanemagus) [inline package.json]: Don't put this in, we'll simply not merge this until `stylelint` has been updated.

File: lib/index.js [hunk 2]
```diff
@@ -46,6 +53,9 @@ export function activate() {
   subscriptions.add(atom.config.observe('linter-stylelint.disableWhenNoConfig', value => {
     disableWhenNoConfig = value;
   }));
+  subscriptions.add(atom.config.observe('linter-stylelint.enableHtmlStyleLinting', value => {
+    enableHtmlStyleLinting = value;
```
Comment (Arcanemagus) [inline lib/index.js]: You need to handle the option being _disabled_ as well, for an example of how `linter-eslint` does this see [here](https://github.com/AtomLinter/linter-eslint/blob/v7.1.3/lib/main.js#L78-L87).

**Comment by Arcanemagus** [comment]
Can you add a spec testing that this is working?

**Comment by Arcanemagus** [comment]
Thanks for getting this started btw!

**Comment by gaurav21r** [comment]
@Arcanemagus Now that @blake-newman has successfully merged his code into https://github.com/stylelint/stylelint/pull/1019 Can we go ahead with this? 

For projects that use Polymer / WebComponents. Scoping `<style>` within the HTML is sometimes the **only** way to go! And `stylelint` linting only `.css` files touches across only a small percentage of styles in our project!

**Comment by Arcanemagus** [comment]
@gaurav21r This still needs a spec, and a released version of `stylelint` incorporating it before this can be merged :wink:.

**Comment by blake-newman** [comment]
@Arcanemagus @gaurav21r Will be updating this in the coming few days.

**Comment by Arcanemagus** [comment]
@blake-newman I'd like to get this released soon, we're you planning on addressing the comments above and updating it for how the final implementation ended up? I can just pull your commit into a new PR and update it there if you don't have time right now :wink:.

**Comment by blake-newman** [comment]
@Arcanemagus All is done :)

File: package.json [hunk 3]
```diff
@@ -36,9 +36,9 @@
     "atom-linter": "^4.6.1",
     "atom-package-deps": "^4.0.1",
     "cosmiconfig": "^1.1.0",
-    "lazy-req": "^1.1.0",
-    "stylelint": "6.3.2",
-    "stylelint-config-standard": "^7.0.0"
+    "deep-assign": "^2.0.0",
+    "stylelint": "^6.3.2",
+    "stylelint-config-standard": "^4.0.1"
```
Comment (1000ch) [inline package.json]: It seems that `master` branch version is correct.

File: package.json [hunk 4]
```diff
@@ -36,9 +36,9 @@
     "atom-linter": "^4.6.1",
     "atom-package-deps": "^4.0.1",
     "cosmiconfig": "^1.1.0",
-    "lazy-req": "^1.1.0",
-    "stylelint": "6.3.2",
-    "stylelint-config-standard": "^7.0.0"
+    "deep-assign": "^2.0.0",
+    "stylelint": "^6.3.2",
```
Comment (Arcanemagus) [inline package.json]: This is specifically pinned so we get notifications for every new release (and can then push a release of `linter-stylelint` out including it).

File: package.json [hunk 5]
```diff
@@ -36,9 +36,9 @@
     "atom-linter": "^4.6.1",
     "atom-package-deps": "^4.0.1",
     "cosmiconfig": "^1.1.0",
-    "lazy-req": "^1.1.0",
```
Comment (Arcanemagus) [inline package.json]: This package will be _completely_ broken without this, why did you remove it?

File: package.json [hunk 6]
```diff
@@ -36,9 +36,9 @@
...
-    "stylelint": "6.3.2",
-    "stylelint-config-standard": "^7.0.0"
+    "deep-assign": "^2.0.0",
```
Comment (Arcanemagus) [inline package.json]: It looks like your package.json changes are simply a reversion to a much older version... `deep-assign` was moved away from since it has many bugs.

**Comment by blake-newman** [comment]
@Arcanemagus looks as if when i rebased, i incorrectly resolved conflicts. All is resolved now.

**Comment by Arcanemagus** [comment]
Okay, looking over the code now :stuck_out_tongue:.

File: spec/linter-stylelint-spec.js [hunk 7]
```diff
@@ -24,6 +27,7 @@ describe('The stylelint provider for Linter', () => {
     atom.workspace.destroyActivePaneItem();
     atom.config.set('linter-stylelint.useStandard', true);
     atom.config.set('linter-stylelint.disableWhenNoConfig', false);
+    atom.config.set('linter-stylelint.enableHtmlLinting', false);
```
Comment (Arcanemagus) [inline spec/linter-stylelint-spec.js]: The default value is false, no need to explicitly set it to false here.

File: spec/linter-stylelint-spec.js [hunk 8]
```diff
@@ -213,6 +217,34 @@ describe('The stylelint provider for Linter', () => {
     });
   });
 
+  describe('works with HTML files and', () => {
+    it('works with stylelint-config-standard', () => {
+      atom.config.set('linter-stylelint.enableHtmlLinting', true);
```
Comment (Arcanemagus) [inline spec/linter-stylelint-spec.js]: You can probably just move this up into the `describe` (and get rid of the second one below).

Comment (blake-newman) [inline spec/linter-stylelint-spec.js]: It's get reset on an onBefore. I placed it in the describe and it took no effect being set to true.

**Comment by Arcanemagus** [comment]
LGTM after those minor nitpicks are addressed. The only thing I can think of that this isn't testing is whether SCSS or Less are linted properly in extracted HTML, but I'm perfectly fine with leaving that testing to `stylelint`.

**Comment by blake-newman** [comment]
@Arcanemagus I tested placing `atom.config.set('linter-stylelint.enableHtmlLinting', true);` once in describe, this took no effect. Removed from `beforeEach`. 

If setting `extractStyleTagsFromHtml` and anything is passed to stylelint then everything is working, i believe the coverage should suffice.

File: spec/linter-stylelint-spec.js (hunk 7 above)
Comment (Arcanemagus) [inline spec/linter-stylelint-spec.js]: I'm still seeing this in the diff?

File: spec/linter-stylelint-spec.js (hunk 8 above)
Comment (Arcanemagus) [inline spec/linter-stylelint-spec.js]: Try adding to a `beforeEach` block within this `describe`?

**Comment by Arcanemagus** [comment]
Commented again on those two parts

**Comment by Arcanemagus** [comment]
Hmmm, I don't think this actually addresses one of my original comments: If HTML linting is enabled, and then disabled, does this still trigger on embedded CSS?

File: lib/index.js [hunk 9]
```diff
@@ -25,12 +25,19 @@ export const config = {
     description: 'Either .stylelintrc or stylelint.config.js',
     type: 'boolean',
     default: false
+  },
+  enableHtmlLinting: {
+    title: 'Enable linting of styles within html',
+    description: 'Turn on linting of your styles tages within html files.',
```
Comment (Arcanemagus) [inline lib/index.js]: "style tags"

**Comment by Arcanemagus** [comment]
`provideLinter` is only called once: When the service is registered with linter. The grammar scopes need to be updated in a manner similar to how linter-eslint [does it](https://github.com/AtomLinter/linter-eslint/blob/v7.2.2/src/main.js#L76-L84).

**Comment by Arcanemagus** [comment]
Merging this and fixing separately as I would like to get this released.

Thanks for putting this together @blake-newman!

**Comment by Arcanemagus** [comment]
Fixes are up in #194.

**Comment by gaurav21r** [comment]
Thanks a TON @Arcanemagus @blake-newman :+1: :+1: :+1: :+1: :100:  / :100:
